packages/@react-aria/list/src/useListItem.ts

```diff
@@ -0,0 +1,204 @@
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {focusSafely, getFocusableTreeWalker} from '@react-aria/focus';
+import {getRowId, listMap} from './utils';
+import {HTMLAttributes, KeyboardEvent as ReactKeyboardEvent, RefObject} from 'react';
+import {isFocusVisible} from '@react-aria/interactions';
+import type {ListState} from '@react-stately/list';
+import {mergeProps} from '@react-aria/utils';
+import {Node as RSNode} from '@react-types/shared';
+import {useLocale} from '@react-aria/i18n';
+import {useSelectableItem} from '@react-aria/selection';
+
+export interface AriaListItemOptions {
+  /** An object representing the list item. Contains all the relevant information that makes up the list row. */
+  node: RSNode<unknown>,
+  /** Whether the list row is contained in a virtual scroller. */
+  isVirtualized?: boolean,
+  /** Whether selection should occur on press up instead of press down. */
+  shouldSelectOnPressUp?: boolean,
+  /** Whether the list item is disabled. */
+  isDisabled?: boolean
+}
+
+export interface ListItemAria {
+  /** Props for the list row element. */
+  rowProps: HTMLAttributes<HTMLElement>,
+  /** Props for the grid cell element within the list row. */
+  gridCellProps: HTMLAttributes<HTMLElement>,
+  /** Whether the row is currently pressed. */
+  isPressed: boolean
+}
+
+/**
+ * Provides the behavior and accessibility implementation for a row in a list.
+ * @param props - Props for the row.
+ * @param state - State of the parent list, as returned by `useListState`.
+ * @param ref - The ref attached to the row element.
+ */
+export function useListItem<T>(props: AriaListItemOptions, state: ListState<T>, ref: RefObject<HTMLElement>): ListItemAria {
+  // Copied from useGridCell + some modifications to make it not so grid specific
+  let {
+    node,
+    isVirtualized,
+    shouldSelectOnPressUp,
+    isDisabled
+  } = props;
+
+  let {direction} = useLocale();
+  let {onAction} = listMap.get(state);
+  let focus = () => {
+    // Don't shift focus to the row if the active element is a element within the row already
+    // (e.g. clicking on a row button)
+    if (!ref.current.contains(document.activeElement)) {
+      focusSafely(ref.current);
+    }
+  };
+
+  let {itemProps, isPressed} = useSelectableItem({
+    selectionManager: state.selectionManager,
+    key: node.key,
+    ref,
+    isVirtualized,
+    shouldSelectOnPressUp,
+    onAction: onAction ? () => onAction(node.key) : undefined,
+    isDisabled,
+    focus
+  });
+
+  let onKeyDown = (e: ReactKeyboardEvent) => {
```

Comment:
Yeah it would be useful for other single column grid like components, I can pull it out into its own hook when we need it. Not sure there is one that exists right now though (TagGroup handles left/right differently), but I maybe forgetting one